packages/app-desktop/gui/NewWindowOrIFrame.tsx

```diff
@@ -0,0 +1,178 @@
+import { createPortal } from 'react-dom';
+import { SecondaryWindowApi } from '../utils/window/types';
+
+// This component uses react-dom's Portals to render its children in a different HTML
+// document. As children are rendered in a different Window/Document, they should avoid
+// referencing the `window` and `document` globals. Instead, HTMLElement.ownerDocument
+// and refs can be used to access the child component's DOM.
+
+export const WindowIdContext = createContext(defaultWindowId);
+
+type OnCloseCallback = ()=> void;
+type OnFocusCallback = ()=> void;
+
+export enum WindowMode {
+	Iframe, NewWindow,
+}
+
+interface Props {
+	// Note: children will be rendered in a different DOM from this node. Avoid using document.* methods
+	// in child components.
+	children: React.ReactNode[]|React.ReactNode;
+	title: string;
+	mode: WindowMode;
+	windowId: string;
+	onClose: OnCloseCallback;
+	onFocus?: OnFocusCallback;
+}
+
+const useDocument = (
+	mode: WindowMode,
+	iframeElement: HTMLIFrameElement|null,
+	onClose: OnCloseCallback,
+) => {
+	const [doc, setDoc] = useState<Document>(null);
+
+	const onCloseRef = useRef(onClose);
+	onCloseRef.current = onClose;
+
+	useEffect(() => {
+		let openedWindow: Window|null = null;
+		const unmounted = false;
+		if (iframeElement) {
+			setDoc(iframeElement?.contentWindow?.document);
+		} else if (mode === WindowMode.NewWindow) {
+			openedWindow = window.open('about:blank');
+			setDoc(openedWindow.document);
+
+			// .onbeforeunload and .onclose events don't seem to fire when closed by a user -- rely on polling
+			// instead:
+			void (async () => {
+				while (!unmounted) {
+					await new Promise<void>(resolve => {
+						shim.setTimeout(() => resolve(), 2000);
+					});
+
+					if (openedWindow?.closed) {
+						onCloseRef.current?.();
+						openedWindow = null;
+						break;
+					}
+				}
+			})();
+		}
+
+		return () => {
+			// Delay: Closing immediately causes Electron to crash
+			setTimeout(() => {
+				if (!openedWindow?.closed) {
+					openedWindow?.close();
+					onCloseRef.current?.();
+					openedWindow = null;
+				}
+			}, 200);
+
+			if (iframeElement && !openedWindow) {
+				onCloseRef.current?.();
+			}
+		};
+	}, [iframeElement, mode]);
+
+	return doc;
+};
+
+type OnSetLoaded = (loaded: boolean)=> void;
+const useDocumentSetup = (doc: Document|null, setLoaded: OnSetLoaded, onFocus?: OnFocusCallback) => {
+	const onFocusRef = useRef(onFocus);
+	onFocusRef.current = onFocus;
+
+	useEffect(() => {
+		if (!doc) return;
+
+		doc.open();
+		doc.write('<!DOCTYPE html><html><head></head><body></body></html>');
+		doc.close();
+
+		const cssUrls = [
+			'style.min.css',
+			'style/icons/style.css',
+			'vendor/lib/@fortawesome/fontawesome-free/css/all.min.css',
+			'vendor/lib/react-datetime/css/react-datetime.css',
+			'vendor/lib/smalltalk/css/smalltalk.css',
+			'vendor/lib/roboto-fontface/css/roboto/roboto-fontface.css',
+			'vendor/lib/codemirror/lib/codemirror.css',
+		];
```

Comment:
Currently, these are hardcoded to match the CSS URLs in the main `index.html` document. I'm currently trying to update this so that `style.min.css` imports the other CSS files dynamically.